src/miette_diagnostic.rs

```diff
@@ -0,0 +1,96 @@
+use std::{
+    error::Error,
+    fmt::{Debug, Display},
+};
+
+use crate::{Diagnostic, Severity};
+
+/// Diagnostic that can be created at runtime.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MietteDiagnostic {
+    /// Displayed diagnostic description
+    pub description: String,
+    /// Unique diagnostic code to look up more information
+    /// about this Diagnostic. Ideally also globally unique, and documented
+    /// in the toplevel crate's documentation for easy searching.
+    /// Rust path format (`foo::bar::baz`) is recommended, but more classic
+    /// codes like `E0123` will work just fine.
+    pub code: Option<String>,
+    /// [`Diagnostic`] severity. Intended to be used by
+    /// [`ReportHandler`](crate::ReportHandler)s to change the way different
+    /// [`Diagnostic`]s are displayed. Defaults to [`Severity::Error`].
+    pub severity: Severity,
+}
+
+impl Display for MietteDiagnostic {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", &self.description)
+    }
+}
+
+impl Error for MietteDiagnostic {}
+
+impl Diagnostic for MietteDiagnostic {
+    fn code<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+        self.code
+            .as_ref()
+            .map(Box::new)
+            .map(|c| c as Box<dyn Display>)
+    }
+
+    fn severity(&self) -> Option<Severity> {
```

Comment:
I would suggest to change `Diagnostic` to just return `Severity`, but this will be breaking change.
Using `Option<Severity>` is somewhat unreasonable, as every diagnostic haa some severity. And we can use `Severity::default()` instead of `None`

Reply:
I think the current optional return type is useful in scenarios where you have an inner error proving must of the `miette` implementation.

For example:
```rust
#[derive(Debug, Error)]
#[error("{inner}")]
pub struct Outer {
  pub inner: Inner
}

impl miette::Diagnostic for Outer {
  fn severity(&self) -> Option<miette::Severity> {
    // I want all outer errors to be warnings by default
    // except if overridden by the inner error.
    self.kind.severity().or(Some(miette::Severity::Warning))
  }
}

#[derive(Debug, Diagnostic, Error)]
pub enum Inner {
  NotThatBad,
  
  #[diagnostic(severity("error"))]
  ReallyBad
}
```

Reply:
Sounds reasonable